inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoRenderer.java

```diff
@@ -108,35 +110,62 @@ public static PdfAnnoModel render(AnnotatorState aState, VDocument aVDoc, String
-            (range.getEnd() < aDocumentText.length() - offset
-                ? offset : aDocumentText.length() - range.getEnd()), annotatedText.length()) ;
-        // remove whitespaces as they are not present in PDFExtract text
-        String cleanAnnotatedText = annotatedText.replaceAll("\\s", "");
-
-        int index = aPdfExtractFile.getStringContent().indexOf(cleanAnnotatedText);
-        if (index < 0) {
-            LOG.error("Could not map exisiting annotation with id " + aVspan.getVid().toString());
-            return null;
-        } else {
-            start = index + preText.replaceAll("\\s", "").length() + 1;
-            end = index + cleanAnnotatedText.length() - postText.replaceAll("\\s", "").length();
-            PdfExtractLine first = aPdfExtractFile.getStringPdfExtractLine(start);
-            PdfExtractLine last = aPdfExtractFile.getStringPdfExtractLine(end);
-            return new Span(aVspan.getVid().getId(), first.getPage(), aColor,
-                cleanAnnotatedText.substring(start - index, end - index).replaceAll("\\s", ""),
-                first.getPosition(), last.getPosition());
+        int docTextLen = aDocumentText.length();
+        int vSpanBegin = aDocumentText.length();
+        int vSpanEnd = -1;
+        // search for begin of the first range and end of the last range and use those
+        for (VRange range : aVspan.getRanges())
+        {
+            if (range.getBegin() < vSpanBegin) vSpanBegin = range.getBegin();
+            if (range.getEnd() > vSpanEnd) vSpanEnd = range.getEnd();
         }
+
+        // use an offset before and after string to find a unique text snippet for an annotation
+        // begin with 0 and increase by 1 until a unique text is found
+        // offset is limited to length of a quarter of the text
+        for (int offset = 0; offset < docTextLen / 4 + 1; offset++)
+        {
+            // subtract offset from vSpanBegin and add offset to vSpanEnd and stay in bounds
+            int offsetBegin = vSpanBegin <= offset ? 0 : vSpanBegin - offset;
+            int offsetEnd = vSpanEnd < docTextLen - offset ? vSpanEnd + offset : docTextLen - 1;
+
+            // get annotated text from aDocumentText with offset before and after it
+            // also remove all whitespaces because they do not exist in PDFExtract text
+            String annotatedText = aDocumentText.substring(offsetBegin, offsetEnd).replaceAll("\\s", "");
+            String offsetBeforeText = aDocumentText.substring(offsetBegin, vSpanBegin).replaceAll("\\s", "");
+            String offsetAfterText = aDocumentText.substring(vSpanEnd, offsetEnd).replaceAll("\\s", "");
+
+            List<Integer> indices =
+                getAllIndices(aPdfExtractFile.getStringContent(), annotatedText);
+            if (indices.size() == 0) {
+                // if there are no matches end search as adding more characters won't help anymore
+                break;
+            } else if (indices.size() == 1) {
```

Comment:
What if the size of indices is larger than 1?

Reply:
Search is continued by increasing offset because there is no unique match found but multiple possible matches.
With increased offset chances are better to find a unique match.